### ISSUE TITLE
Is there a runtime alternative to `quote!` that interpolates a `&str`?

### ISSUE THREAD

**Issue by mitchmindtree** [dtolnay/quote#100]
Hi @dtolnay - firstly, thank you so much for your work on serde and the procedural macro system! I think Rust would be a long shot from what it is today without your work :)

Can you recommend an approach for getting `quote!`-esque behaviour at runtime by interpolating a `&str`? I'm attempting to develop an API where a user can store a rust expression in the form of a `&str` where the expression might contain placeholders like `#foo` or `#bar` (similarly to `quote!`) which can later be replaced during a code generation phase with other expressions. The end result would then be a `syn::Expr` with the inner `#foo` and `#bar` replaced with the necessary sub expressions.

**TL;DR** I'm looking for a way to go from `&str` -> `syn::Expr` but with some runtime "quasi-quoting" done in between.

Any advice would be appreciated before I dive into writing my own amateur solution!

**Comment by dtolnay** [comment]
I don't know of an existing implementation of this. I would approach it by:

1. Parsing the input string into a proc_macro2::TokenStream;
2. Traversing the TokenStream to find `#` followed by ident and replacing them with a proc_macro2::Group token whose delimiter is proc_macro2::Delimiter::None and which contains the tokens to be inserted;
3. Parsing the TokenStream to syn::Expr.

The Group token is to make sure that something like `#freq.sin()` with `freq = f1 + f2` behaves with the correct precedence.

**Comment by mitchmindtree** [comment]
I had a stab last night and happened to end up pretty close to your approach - the `Group` token and precedence edge-case is something I hadn't considered though and will certainly apply.

Thanks for the tips!
